### DISCUSSION TITLE
Typing an `and` helper

### DISCUSSION THREAD

**Discussion by swastik** [typed-ember/glint#547] [Q&A] [answered]
I'm trying to introduce Glint to our codebase. We use [`ember-truth-helpers`](https://github.com/jmurphyau/ember-truth-helpers) a fair bit. I'm trying to understand how to type `and` such that we can write code like this:

```ts
interface Args {
  onArg1?: () => unknown;
  onArg2?: () => unknown;
}
```

```hbs
{{#if (and @onArg1 @onArg2)}}
  <button {{on "click" @onArg1}}></button>
{{/if}}
```

With that, I'm seeing an error like:

<img width="950" alt="CleanShot 2023-03-24 at 08 08 06 png" src="https://user-images.githubusercontent.com/1569205/227461604-939d2fc9-5997-4bd9-bc3f-8a401fc0d97f.png">

----

Nesting the `if`s works. I can also do other things (e.g. define `onArg1` on the component to ensure it's always defined), so it's not blocking me, but I'm wondering if there's a way to actually type it in a way that `onArg1` gets typed correctly in that block! 🙇🏼‍♂️

**Comment by dfreeman** [comment, marked as answer]
Unfortunately there's no way within TypeScript's type system to express how `&&` works; the logical and comparison operators' behaviors are all essentially "baked in" to the language. Similarly, within Glint we have a notion of [special forms](https://github.com/typed-ember/glint/pull/453) to give special meaning to things like `{{if}}` and `{{yield}}` outside of what would normally be expressible with helper types.

As of https://github.com/typed-ember/glint/pull/510, we now support special forms for logical operators like `&&`, but they aren't actually enabled for any helpers out of the box, for two reasons:
 - Until [RFC 562](https://github.com/cibernox/rfcs/blob/add-logical-operators-to-templates/text/0000-add-logical-operators.md) lands, we have no way of knowing that an arbitrary global `{{and}}` helper actually behaves like JS `&&`
 - As currently specified by RFC 562, the notion of "truthiness" in templates is slightly different than in vanilla JS/TS. There's been [some lengthy discussion](https://github.com/snewcomer/glimmer-vm/pull/3#pullrequestreview-634164135) over whether the semantics should be aligned before the implementation lands.

In the meantime, though, it's possible to opt in to treating `{{and}}` like `&&` (and `{{or}}` like `||`, etc) despite the two points above by setting `additionalSpecialForms` config for `@glint/environment-ember-loose` and/or `@glint/environment-ember-template-imports`:

```js
  "glint": {
    "environment": {
      "ember-loose": {
        "additionalSpecialForms": {
          "globals": {
            "and": "&&",
            "or": "||",
            "not": "!"
          }
        }
      },
      "ember-template-imports": {
        "additionalSpecialForms": {
          "imports": {
            "ember-truth-helpers/helpers/and": { "default": "&&" },
            "ember-truth-helpers/helpers/or": { "default": "||" },
            "ember-truth-helpers/helpers/not": { "default": "!" }
          }
        }
      }
    }
  }
```

We currently don't cover this in our formal documentation anywhere since the semantics of RFC 562 are still in flux, but those options are pretty stable at this point, and it might make sense for us to document them under "Advanced Configuration" or something. In the meantime we can at least point folks to this discussion thread when the question comes up 🙂

**Reply by swastik** [reply]
That's very helpful, thank you!
